### ISSUE TITLE
Orthogonalize the ways that the target of PMIx_Get is determined across different realms.

### ISSUE THREAD

**Issue by dsolt** [pmix/pmix-standard#370]
### Overview

As our working group looked at the presentation of different data realms, we noted that there are some ways that targets within a data realm is specified could be made more uniform.   In some cases, a really elegant solution would break backward compatibility, but there are also other minor changes that could be made that would enhance the current functionality without breaking it. 

Some background:   There are 5 data realms:  application, job (namespace), session, process and node.   Some keys are only defined for a single realm.   Some keys are defined for multiple realms.  Keys that are defined for multiple realms have a default realm.   To indicate that a realm other than the default should be selected you will usually use of the "realm selection attributes":  PMIX_APP_INFO, PMIX_JOB_INFO, PMIX_SESSION_INFO, PMIX_PROC_INFO and PMIX_NODE_INFO.  

There are exceptions however and the presentation in the standard doesn't actually say that you can use these 5 attributes to always select the realms they refer to.   

For example, currently there are no keys which apply to the process realm and another realm.  Likewise for the node realm.  Therefore all keys that apply to the process realm default to that realm and similarly for node realm. This means that PMIX_PROC_INFO and PMIX_NODE_INFO aren't something a user would currently need to use.   For the purpose of presentation and forward compatibility, I think it is clearer to have a uniform presentation of how realm selection is done and just consistently say that you use of of the 5 realm-selection attributes to force the selection of the realm you desire.  That is still a true statement and makes things easier to present.  Also currently we state that PMIX_PROC_INFO isn't something you would need to specify, but for PMIX_NODE_INFO we don't say that it isn't needed.  So I've already made this change in an upcoming PR from the IAWG, but its part of a large set of changes, so it will be while.

The text also currently implies that using a proc.rank of PMIX_RANK_WILDCARD indicates that you want the job realm:

`Job-related information is retrieved by including the namespace of the target job and a rank of PMIX_RANK_WILDCARD in the proc argument passed to PMIX-Get`

I'm actually pretty certain this is not what we want to say, so in the IAWG update to this chapter (coming soon), we have changed this statement to:

`Job-related information can be retrieved by requesting a key which defaults to the job realm or by including the PMIX_JOB_INFO attribute in the info array passed to PMIx_Get. For job-related keys the target job is specified by setting the namespace of the target job in the proc argument and specifying a rank of PMIX_RANK_WILDCARD in the proc argument passed to PMIx_Get.`

This RFC is more about how targets are specified once the target realm is determined.  For these we have to look at each individual realm:

**Application realm**

Here the target application is selected in order as:

1.  PMIX_APPNUM of proc.nspace
2. PMIX_APPNUM of calling process if proc is NULL
3. Application of process specified in proc
4. 0 if proc.rank is WILDCARD and proc.nspace is not the caller's namespace

This is a lot of options and since we can't easily remove any choices, I don't have any recommendations for the application realm.  I wish the only choices were 1 and 3 because I think we can make all the realms have primary and secondary targets that are simple and similarly specified across realms, but for now lets assume we need to keep these and move on.

** process realm **
 
I believe the order here is:

1. The process specified in proc relative to the session specified by PMIX_SESSION_ID
2. The process specified in proc relative to the caller's session id.

My recommendation here is that we allow user's to specify a process using the PMIX_PROCID attribute.  In fact, I kind of wish that PMIx_Get didn't even take a proc_t structure as a parameter, but too late for that.   When this was created, nobody ever dreamed we would have 5 realms, so the PMIX_Get API is focused on the process realm.  Anyway, we could have the following order:

1.  The process specified in PMIX_PROCID
2. The process specified in proc relative to the session specified by PMIX_SESSION_ID.
3. The process specified in proc relative to the caller's session id.

It may be tempting to change this to:

1. The process specified in proc relative to the session specified by PMIX_SESSION_ID.
2. The process specified in proc relative to the caller's session id.
3.  The process specified in PMIX_PROCID

to make sure existing code doesn't break if an existing client is unintentionally passing a (possibly garbage) PMIX_PROCID into a PMIx_Get call that targets the process realm and it is currently being ignored.  However, it may be helpful to think about the other realms before determining which order is best.

** job realm **

The job realm appears to have the following selection priorities:

1. The job/namespace specified in proc.nspace relative to the session specified by PMIX_SESSION_ID.
2. The job/namspace specified in proc.nspace relative to the caller's session id.

Here again, I wonder if for completeness it would make sense to allow PMIX_NSPACE to be used to specify the job/namespace instead of proc.nspace?   I'm not sure where in the priority order to place this as well.   Perhaps it should be an error if you specified both a proc and PMIX_NSPACE?  

**  session realm **

1. session specified by PMIX_SESSION_ID
2. Session id of proc.nspace

No changes suggested here, though I will point out that the attribute has priority over the proc value, which is something we might want to consider when considering adding support for PMIX_NSPACE and PMIX_PROC

** node realm **

1. PMIX_NODEID or PMIX_HOSTNAME (I assume an error if both are specified and they don't refer to the same host?)
2. node id of process specified in proc

Again, another indication that attributes take precedence.  

### Motivation

The motivation for adding PMIX_NSPACE and PMIX_PROCID as selection mechanisms to PMIx_Get is to make the selection process more orthogonal and possibly allow code to more easily generate PMIX_Get calls by using and manipulating only attributes without having to deal with proc information.   

### Discussion Items

Is this worth doing?   I have mixed feelings about it, but I didn't want to lose the idea as we discussed it in our WG, so I just wanted to get it all down here.

**Comment by rhc54** [comment]
> there are no keys which apply to the process realm and another realm. Likewise for the node realm

I don't think that is true - definitely not true for the node realm as we have to use `PMIX_NODE_INFO` in a number of places to ensure we get the right values. I'll have to check about the process realm, but I'm pretty sure there are one or two keys that overlap.

> This RFC is more about how targets are specified once the target realm is determined.

Afraid I don't quite grok where you folks are trying to go with this one. I imagine we can try to provide clearer statements about realm specifications, but some of the logic expressed above seems off from what actually happens. Maybe I've just been using these functions for too long to still be confused, but it feels like we're trying to "overdefine" things to the point where we confuse ourselves.

> Is this worth doing? I have mixed feelings about it, but I didn't want to lose the idea as we discussed it in our WG, so I just wanted to get it all down here.

I'm not entirely clear on what you are proposing to do, so it is hard to advise on whether it is worth doing. If I understand your motivation statement, it sounds like you are mostly looking for a way to not pass the procID to `PMIx_Get`? Trying to circumvent the `proc` parameter to `PMIx_Get` is a quick way to get into trouble, and I would strongly advise against it. If it bothers someone that much, they can use `PMIx_Query_info` instead - though there is some performance penalty in doing so, and you do have to be careful in how you construct the query objects.

**Comment by dsolt** [comment]
Yes, I see now that PMIX_MAX_PROCS is an example that applies to the node realm as well as other realms, though that doesn't change anything about what is being proposed here. 

The proposal here is just whether to allow PMIX_NSPACE and PMIX_PROCID to specify the target within the job and process realms respectively.  

I guess what I really wish is that since the idea of realms has expanded since PMI_Get() first came on the scene, I kind of wish that PMIX_Get didn't take a proc at all, just a key and a list of attributes.   Then each realm would behave the same way. We wouldn't have all these issues about if proc is NULL or if proc.rank is wildcard or if PMIX_SESSION_ID is set, etc.  Everything would just be specified in info's.  Truth is we don't have that and maybe at this point trying to move the attribute interface in that direction just confuses things further?

**Comment by rhc54** [comment]
I'm not sure it confuses things, but I'm also not sure it is really necessary. If you look at how this is used, the overwhelming majority of the time someone either is requesting info (a) about the job (which means they specify proc = {namespace, wildcard}) or (b) about a specific proc. So passing a procID is the simplest method for nearly all uses. We made one extension here by allowing the caller to use `NULL` in place of their own procID so they wouldn't have to retain a global `pmix_proc_t` value.

Where things get a little less clear are:

(a) the legacy scenarios where the keys are global and not associated with a proc. We had already used the procID=NULL case, so the only option would be to either define a "this is a global key" attribute or indicate that the requested info was a global key by specifying the use of an `INVALID` rank in the procID. We chose the latter - could go either way, really.

(b) when someone requests info that is not associated with the job or a proc (e.g., about the node). This is where we perhaps got a little too flexible. We could have just insisted that the caller leave the procID as `NULL` or perhaps some invalid value and then pass the identifying info in the attributes - e.g., the name of the node whose info is being requested. Instead, we tried to allow for more general usage by letting the caller pass a valid procID and request node-realm info about "the node where this proc is executing". Perhaps it would have been less confusing if we had forced the user to first request the ID of the node where procID is executing, and then request node-realm info about that node - i.e., make them execute the request in two distinct steps instead of aggregating it in one. Not sure of the best choice here.

Ultimately, there are multple possible solutions, none of which are perfect. Probably the best one can do is provide examples and explanation. 🤷‍♂️
